fix: add favicon fetching hook

NicolasMassart opened the pull request:
## Description

- Replace Faviconkit url with a hook that returns favicon url
- make it work also for SVG files and handle fallback
- create a favicon utils file for fetching in HTML and dealing with origin and urls
- create wrapper to use hook in class components
- update reducer to store favicon url in the state as a cache (max 100)
- create new unit tests and update impacted existing ones
- also added a list of files to ignore by the `yarn format` tool so that we can run it in our branches without having tons of unexpected changes on unrelated files

## Issue

fixes https://github.com/MetaMask/mobile-planning/issues/1161

## E2E tests

🟢 [Pipeline build #9783](https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/e913c768-18fc-4e29-a9da-33530bd4404f?tab=workflows)

## Checklist

* [x] There is a related GitHub issue
* [x] Tests are included if applicable
* [x] Any added code is fully documented

## Test Scenarios

### Scenario: Website favicon should appear next to an autocompleted result when searching for websites with a favicon

Case: This is a normal use case of autocompleting a dapp that exists in [the dapps list](https://github.com/MetaMask/metamask-mobile/blob/main/app/util/dapp-url-list.js)

- Given I am on the browser tab
- And I have the default Metamask home page open (home.metamask.io)
- When I type `open` in the search bar
- Then a ![image](https://github.com/MetaMask/metamask-mobile/assets/4677568/e9b38931-872e-44a0-9814-dc3b30722c5b) favicon should appear to the left of the autocompleted option `opensea.io`
- <img width="188" alt="OpenSea logo" src="https://github.com/MetaMask/metamask-mobile/assets/4677568/6e3855b8-74ad-4666-8a2d-f7ac6a244cdb">
  
### Scenario: Fallback letter should appear next to an autocompleted result when searching for websites without a favicon

Case: this is an error case where the dapp that exists in [the dapps list](https://github.com/MetaMask/metamask-mobile/blob/main/app/util/dapp-url-list.js) either doesn't have a favicon or favicon is not available.

- Given I am on the browser tab
- And I have the default Metamask home page open (home.metamask.io)
- When I type `metamask` in the search bar
- Then a <img width="20" alt="M" src="https://github.com/MetaMask/metamask-mobile/assets/4677568/6285b513-9631-4416-802c-b65db0aa1962"> favicon should appear to the left of the autocompleted option `Metamask Ether Faucet (Ropsten)` as the site doesn't exist anymore.
- <img width="186" alt="image" src="https://github.com/MetaMask/metamask-mobile/assets/4677568/f843c851-3a65-47e1-acb9-a0974b5bf5bc">

### Scenario: Website favicon should appear on the browser tabs list

Case: This is a normal use case for websites that have a favicon

- Given I am on the browser tab
- And I have the default Metamask home page open (home.metamask.io)
- When I type `app.uniswap.org` in the search bar and validate
- Then the Uniswap website opens
- <img width="197" alt="image" src="https://github.com/MetaMask/metamask-mobile/assets/4677568/1df4a0b3-1eaa-4964-9a6e-c06f1dc5baf0">
- When I touch the bottom bar tab action <img width="32" alt="image" src="https://github.com/MetaMask/metamask-mobile/assets/4677568/5b8f3b01-27c7-4ff7-a963-3b868909fe7b">
- Then the tabs list appears showing one tab for Uniswap with the website favicon
- <img width="153" alt="image" src="https://github.com/MetaMask/metamask-mobile/assets/4677568/b3130211-1fcf-4f3c-bf3a-b325cbf6beac">

### Scenario: Website favicon should appear in connect and permission modals

Case: This is a normal use case for websites that have a favicon

- Given I am on the browser tab
- And I have the default Metamask home page open (home.metamask.io)
- And the test dapp website is loaded
- <img width="156" alt="image" src="https://github.com/MetaMask/metamask-mobile/assets/4677568/8aa67eca-e31c-4fc0-a80b-4014620a1f6a">
- When I touch "connect" in "Basic Actions" on the dapp
- Then a connect modal appears showing the dapp favicon and the URL
- <img width="155" alt="image" src="https://github.com/MetaMask/metamask-mobile/assets/4677568/195db66b-88e9-43b4-854d-a954bd604296">
- When I touch "Connect"
- Then the modal hides
- And dapp "Connect" button is disabled and top right account button <img width="37" alt="image" src="https://github.com/MetaMask/metamask-mobile/assets/4677568/f7f7bcd3-5926-485a-abaa-727bd550539b"> appears
- When I touch top right account button
- Then the connected accounts modal appears with the favicon and the domain name
- <img width="154" alt="image" src="https://github.com/MetaMask/metamask-mobile/assets/4677568/2eaa9ff0-853b-4652-b0c0-7a2ebdef81bc">
- When I touch permission link
- Then Permission modal appears with favicon and url of the dapp
- <img width="152" alt="image" src="https://github.com/MetaMask/metamask-mobile/assets/4677568/44302f28-7d95-4520-a2a5-702eb860dea7">

### Scenario: Website favicon should appear in transaction modal

Case: This is a normal use case for websites that have a favicon

- Given I am on the test dapp
- And I connected my account
- When I touch the "send legacy transaction" button
- Then the transaction modal displays the dapp favicon
- <img width="157" alt="image" src="https://github.com/MetaMask/metamask-mobile/assets/4677568/9507d2c7-6e44-47ef-b4ea-996da2549a30">

### Scenario: Website favicon should appear in signing modal

Case: This is a normal use case for websites that have a favicon

- Given I am on the test dapp
- And I connected my account
- When I touch the "Sign" button in Personal sign section
- Then the signing modal displays the dapp favicon
- <img width="150" alt="image" src="https://github.com/MetaMask/metamask-mobile/assets/4677568/7a6fe980-194e-42c4-a332-7583b7433f29">

### Scenario: Website favicon should appear in favourites on home page

**NOTE: this scenario can not be successful until [the home.metamask.io page is also updated](https://github.com/MetaMask/dapps/pull/188)**

Case: This is a normal use case for websites that have a favicon

- Given I added a website with a favicon to my favourites
- <img width="156" alt="Screenshot 2023-09-26 at 17 19 29" src="https://github.com/MetaMask/metamask-mobile/assets/4677568/33bfeb59-e65a-4763-b55e-f3cf0cf0279d"><img width="155" alt="image" src="https://github.com/MetaMask/metamask-mobile/assets/4677568/46a3b9e8-87c2-4406-9474-1f42b92506d1">
- When I touch the <img width="32" alt="home" src="https://github.com/MetaMask/metamask-mobile/assets/4677568/89571924-56a4-43e4-810a-72f538d02ef1"> button in the browser bar
- Then the home page should display my favourited site in the list of favourites
- And the favicon should be the same as the browser tabs list one

### Scenario: Website favicon should appear in connect modal when deeplinking from WalletConnect

Case: This is a normal use case for websites that have a favicon

- Given I am on a dapp (ie. app.uniswap.org) on my native mobile browser
- And I connected my account with the WalletConnect feature
- And I touch the "connect with Metamasks" button
- And the MM app opens with the wallet connect modal for this dapp
- Then the wallet connect modal on MM app displays the dapp favicon

https://github.com/MetaMask/metamask-mobile/assets/4677568/9396ccd4-6768-45f4-a903-c631dcd30f45

### Scenario: Website favicon should appear in connect modal when scanning QRCode from SDK enabled dapp

Case: This is a normal use case for websites that use MM SDK

- Given I am on a SDK enabled dapp (ie. [SDK test dapp](https://c0f4f41c-2f55-4863-921b-sdk-docs.github.io/test-dapp/)) on desktop browser
- And I connected my account with the connect button
- <img width="200" alt="image" src="https://github.com/MetaMask/metamask-mobile/assets/4677568/0d42ca29-794f-46bc-accc-ca34ca25f21a"> (this QRCode screenshot can be directly used for testing)
- And I scan the QRcode with Metamask Mobile app
- And the MM app opens with the SDK connect modal for this dapp
- Then the SDK connect modal on MM app displays a QRCode Favicon

https://github.com/MetaMask/metamask-mobile/assets/4677568/b246f8c7-8841-48d0-a5ea-707827e0ce59

NicolasMassart committed:
add favicon fetching hook

NicolasMassart committed:
Merge branch 'main' into fix/1161_favicon_download

NicolasMassart committed:
making the code more readable

NicolasMassart committed:
clean code and add cache

NicolasMassart committed:
Merge branch 'fix/1161_favicon_download' of github.com:MetaMask/metamask-mobile into fix/1161_favicon_download

NicolasMassart committed:
fix extension return

NicolasMassart committed:
move favicon html parsing to utils

NicolasMassart committed:
remove comments

NicolasMassart committed:
fix unit tests

NicolasMassart committed:
Merge branch 'main' into fix/1161_favicon_download

NicolasMassart committed:
update xmldom

NicolasMassart committed:
update unit test and fix useEffect deps

NicolasMassart committed:
caching favicon url

NicolasMassart committed:
Merge branch 'main' into fix/1161_favicon_download

NicolasMassart commented on app/components/hooks/useFavicon/useFavicon.ts:
```diff
@@ -0,0 +1,32 @@
+    useState<ImageSourcePropType>(EMPTY_FAVICON_URI);
+
+  useEffect(() => {
+    (async () => {
```
error handling is missing in the async function inside useEffect. Might want to add a try/catch block to handle any potential errors during the favicon fetching and caching process.

NicolasMassart commented on app/util/favicon/index.test.ts:
```diff
@@ -0,0 +1,151 @@
+  /**
+   * should be able to find favicon from html links collection
+   */
+  it('should return favicon from html', async () => {
```
It could be beneficial to add additional test cases for possible error scenarios, for instance, when fetch fails, or the HTML doesn't contain a link to a favicon.

hesterbruikman commented:
[PR](https://github.com/MetaMask/dapps/pull/188) for home.metamask.io 

@NicolasMassart @cortisiko it appears the 188 cannot be tested unless merged. I'll ask to hold of to next sprint so that both PR's can be merged and tested in parallel. Please let me know if that doesn't make sense.

NicolasMassart committed:
Merge branch 'main' into fix/1161_favicon_download

# Conflicts:
#	app/core/AppConstants.ts

NicolasMassart committed:
adding hook to class controlers

NicolasMassart committed:
update tests

NicolasMassart committed:
Merge branch 'main' into fix/1161_favicon_download

NicolasMassart committed:
add error handling

NicolasMassart committed:
update snapshots

NicolasMassart committed:
Merge branch 'fix/1161_favicon_download' of github.com:MetaMask/metamask-mobile into fix/1161_favicon_download

NicolasMassart committed:
fix the test

NicolasMassart committed:
add reducer test

NicolasMassart committed:
update unit test

NicolasMassart committed:
convert the tests back to JS to match reducer

Ideally reducer should be refactored to TS at some point

NicolasMassart committed:
fix bookmark

NicolasMassart committed:
Merge branch 'main' into fix/1161_favicon_download

Cal-L commented on app/util/favicon/index.ts:
```diff
@@ -0,0 +1,152 @@
+};
+
+const originToUrl = (origin: string) => {
+  if (origin) {
```
You can use `is-url` to check if it's a valid URL.

NicolasMassart replied:
The issue with `isUrl` is that [it first checks for the protocol](https://github.com/segmentio/is-url/blob/master/index.js#L31-L34) and so if the string to test is just `domain.tld` and not `someting://domain.tld` then it will return false.

The origin as it's stored in state is usually not a URL but only the domain name.
When the origin is not a URL, adding the protocol allows to use it as input for `new URL(string)` with only the risk that in the end the address doesn't respond or returns an error and the favicon can handle it by displaying a default image. in case the string doesn't even have a correct form, trying to creatr an URL with it will throw an error that we catch and after which we return undefined.

Ideally the origin stored in our state should be the full URL of the dapp with path and then we only pick the parts we need for display, but currently we discard protocol and path and the stored origin is just the host string. With typing, we could enforce the use of the URL type for origin, but I believe it's out of the scope of this PR as it impacts the origin handling and UI a lot, not only favicon.

Tell me if you think we should create an issue for this and of course if you still think I should find a way to test the validity of the origin differently.

Cal-L replied:
Ah ok. Sounds like it's handled properly down the line so let's leave it as is for now.

Cal-L commented on app/util/favicon/index.ts:
```diff
@@ -0,0 +1,152 @@
+    try {
+      const originWithProtocol = origin.startsWith('http')
+        ? origin
+        : `https://${origin}`;
```
Is there a reason why origin would never have a protocol?

Cal-L replied:
The URL validation should probably also happen after any protocol has been prepended

Cal-L commented on app/util/favicon/index.ts:
```diff
@@ -0,0 +1,152 @@
+  }
+};
+
+const originToCacheKey = (origin: string) => {
```
Might be better to rename this `originToHost` so it's more descriptive.

NicolasMassart committed:
Merge branch 'main' into fix/1161_favicon_download

NicolasMassart commented on app/util/favicon/index.ts:
```diff
@@ -0,0 +1,152 @@
+
+  if (links && links.length > 0 && origin) {
+    // use lodash forEach as the collection require iteration to prevent named items to be returned
+    forEach(links, (link) => {
```
Updated to use Array.every() function

NicolasMassart committed:
add and fix TSDOC comment and code

- use is-url
- fix types
- remove use of lodash foreach
- use URL instead of strings when possible
- rename originToCacheKey function

NicolasMassart committed:
Merge branch 'main' into fix/1161_favicon_download

Cal-L reviewed (dismissed):
Code LGTM!

Cal-L commented on app/util/favicon/index.ts:
```diff
@@ -0,0 +1,155 @@
+ * @param originUrl - the origin used as cache key
+ * @param faviconUrl - the stored favicon url
+ */
+export const cacheFavicon = async (originUrl: string, faviconUrl: string) => {
```
Remove both async from function and await from Logger to remove the need for a try catch.

NicolasMassart replied:
Catching the error allows me to prevent it to throw out of the scope and let the process continue anyway as we have a favicon fallback.
But still log it in case of.
When logging, i need the error to add to the log, so i need the catch to get this error.
Maybe logging is not so necessary and I could remove the log completely, but then I still have the need of the catch and having nothing, not even a log in the catch is not the best imo.

Cal-L commented on app/util/favicon/index.ts:
```diff
@@ -0,0 +1,155 @@
+const originToHost = (origin: string) => {
+  const normalisedOrigin = originToUrl(origin);
+  if (normalisedOrigin) {
+    try {
```
No need for try catch

Cal-L commented on app/util/favicon/index.ts:
```diff
@@ -0,0 +1,155 @@
+ * @param faviconUrl - the stored favicon url
+ */
+export const cacheFavicon = async (originUrl: string, faviconUrl: string) => {
+  try {
```
No need for try catch. Remove await from Logger

Cal-L commented on app/util/favicon/index.ts:
```diff
@@ -0,0 +1,155 @@
+  const cacheKey = originToHost(originUrl);
+  if (!cacheKey) return;
+  let faviconUrl = null;
+  try {
```
No need for async, try catch, and remove await from Logger

NicolasMassart committed:
Remove useless try catch

Replace toString by direct use of URL
fix empty url warning on image
and update deprecated substr

NicolasMassart committed:
Merge branch 'fix/1161_favicon_download' of github.com:MetaMask/metamask-mobile into fix/1161_favicon_download

NicolasMassart committed:
add SVG support

NicolasMassart committed:
Merge branch 'main' into fix/1161_favicon_download

# Conflicts:
#	yarn.lock

NicolasMassart committed:
update packages

NicolasMassart committed:
update for fallback favicon

actual favicon pulled on rerendered step because of useEffect
this has no impact on the result of this component test

NicolasMassart commented on .storybook/storybook.requires.js:
```diff
@@ -6,35 +6,35 @@ import {
   addParameters,
   addArgsEnhancer,
   clearDecorators,
-} from "@storybook/react-native";
```
.storybook/storybook.requires.js was formated after merge with main, not part of this changes... ignore for review.

NicolasMassart committed:
revert format on a file that should have been ignored

and add ignore for all these files that are not formatted
and that are still taken into account when running
yarn format manually

NicolasMassart committed:
Merge branch 'main' into fix/1161_favicon_download

Cal-L reviewed (dismissed):
LGTM

cortisiko committed:
Merge branch 'main' into fix/1161_favicon_download

NicolasMassart commented:
As discussed with @cortisiko add the state migration for people updating app who already have bookmarked dapps.

NicolasMassart committed:
update favicon for walletconnect deep linking

NicolasMassart committed:
Merge branch 'main' into fix/1161_favicon_download

NicolasMassart committed:
update unit test

NicolasMassart committed:
Merge branch 'main' into fix/1161_favicon_download

Cal-L reviewed (dismissed):
Code LGTM.

NicolasMassart committed:
Merge branch 'main' into fix/1161_favicon_download

Cal-L commented:
Added Do not merge label. Should be part of 7.11.0

tommasini committed:
fix svg image size when image on the website is an svg

tommasini committed:
added uniswap.org to the dapp url list, fixed the icon crash

tommasini committed:
merge main and solve conflicts

legobeat committed:
run `yarn deduplicate`

tommasini committed:
update uniswap.org url to app.uniswap.org on dapp list urls

cortisiko reviewed (approved):
✅ ✅ ✅ ✅ ✅ ✅

tommasini committed:
Merge branch 'main' into fix/1161_favicon_download
